### ISSUE TITLE
Failed device assertion results in memory access fault

### ISSUE THREAD

**Issue by Maetveis** [ROCm-Developer-Tools/HIP#2482]
The following simple code when compiled using `hipcc -O0 main.cpp` triggers a memory access fault and does not print the failed assertion.

```cpp
#include <hip/hip_runtime.h>

#include <cassert>

__global__ void f() {
	assert(false);
	//printf("%d\n", 0);
}

int main(){
	hipLaunchKernelGGL(f, dim3(1), dim3(1), 0, 0);
	hipDeviceSynchronize();
}
```

Result when running on ubuntu 18.04 with gfx906 (aka MI50) using rocm 5.0:
```
Memory access fault by GPU node-4 (Agent handle: 0x1c93d70) on address (nil). Reason: Page not present or supervisor privilege.
Aborted (core dumped)
```

Expected result:
```
main.cpp:6: void f(): Device-side assertion `false' failed.
:0:rocdevice.cpp            :2603: 2338244645597 us: 305  : [tid:0x7f194a44e700] Device::callbackQueue aborting with error : HSA_STATUS_ERROR_EXCEPTION: An HSAIL operation resulted in a hardware exception. code: 0x1016
Aborted (core dumped)
```

If the kernel contains a call to `printf` (even if it is never executed) then the assertion is correctly printed.

**Comment by Maetveis** [comment]
Additional info: when the printf is present then the hidden kernel argument is added to the kernel metadata, but otherwise it isn't.

```
- .address_space:  global
        .offset:         24
        .size:           8
        .value_kind:     hidden_hostcall_buffer
```        
With a bit of digging through llvm, this might be fixed by a recent commit: https://reviews.llvm.org/D119216.

**Comment by Maetveis** [comment]
This was fixed in rocm 5.0.1.
